Refactor

texastoland opened the pull request:
## Proposed changes

1. Abstracts `Proxy` boilerplate
2. Combines `Integers` using default `startIndex`
3. Reduces runtime complexity using `Map`
4. Tweaks naming for unqualified import
5. Adopts EcmaScript module syntax
6. Adds JSDoc annotations to generate `.d.ts`
7. Formatted with Prettier

## Remaining changes if accepted

- [x] Add Prettier config and script (f9130c36116e10093bff16063e5f56b6e8ac75d6)
- [x] [Add `tsconfig.json`](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html) (d04cde59feb283b32e4a8a24d06ca3cfd999dfdb)
- [x] Generate `*.d.ts` from JSDoc (also d04cde59feb283b32e4a8a24d06ca3cfd999dfdb)
- [x] Deprecate previous names (65fdac18b1e210426d280ccda29e264063e12f7b)
- [x] Update tests to use EcmaScript modules (1289066425359b2ef2c2fefa7fea5ec02893dbe0)
- [x] Add `ActionCreators` in `extras` module (4e327765629f2f3907fe6a196611ea7a91274877)
- [x] Add `apply` trap (originally 7b7df53cbd5d179dcaac33b3e31dcc2991bdcf46)
- [ ] Add more tests
- [ ] Add colors in `extras` module
- [ ] Update changelog

## Followup PR

- [ ] Update readme
- [ ] Add custom use cases in examples section

chasefleming commented:
@texastoland This is awesome! I definitely would love to merge this and the remaining changes you mentioned. One note is I think it would be nice to still keep a way of doing integers without having to execute a function like the other styles -- unless someone wants explicitly set a start index -- in which case I don't see a way around it there. Thanks for adding these updates! Excited to see the rest 😀

texastoland commented:
Great thanks! I'll finish tomorrow then. What do you think about:

```js
const Integers = Counter()
```

I realized it's flexible enough for [Redux actions](https://redux-toolkit.js.org/api/createaction/) too:

```js
// partially copied from the official implementation
const ActionCreators = (namespace = "") =>
  MemoEnumOf((name) => {
    const type = namespace ? `${namespace}/${name}` : name
    const actionCreator = (payload, { meta, error = false }) => ({
      type,
      payload,
      meta,
      error,
    })
    actionCreator.type = type
    actionCreator.toString = () => type
    actionCreator.match = (action) => type === action.type
    return actionCreator
  })
```

Usage:

```js
const { increment } = ActionCreators("counter")
// somewhere else

dispatch(increment(step))
// dispatch({ type: "counter/increment", payload: step })

// or simpler yet
const { decrement } = ActionCreators()

dispatch(decrement())
// dispatch({ type: "decrement" })
```

chasefleming commented:
@texastoland Oh wow that's a really cool use case! Saves a lot of lines of code. What do you think about making an examples folder where we put that? Then we could link it on the readme and any other use cases that pop up?

By the way, regarding the naming changes I'm not strongly tied to the original so I'm cool with updating them if you think they are better, but we should probably keep the existing functionality by referencing the new and add a deprecation warning with `console.warn` until they are removed in a later version.

chasefleming commented:
Also, just wanted to say thanks again for your thoughtfulness on these @texastoland ...these are really creative ideas!

mtsknn commented on index.js:
```diff
@@ -1,31 +1,29 @@
+
+const Integers = (startIndex = 0) => {
+  const indexOfName = new Map()
+  return Values((name) => {
+    indexOfName.set(name, indexOfName.size + startIndex)
+    return indexOfName.get(name)
   })
```
Golfing: `Map.set()` returns the `Map` object, so you could chain `.set()` and `.get()` and use implicit return. 🙂

```suggestion
  return Values((name) => 
    indexOfName.set(name, indexOfName.size + startIndex).get(name)
  )
```

texastoland replied:
Thanks I'll do that now ✌🏼

mtsknn commented on index.js:
```diff
@@ -1,31 +1,28 @@
-  get: (_, name) => name
-})
+/**
+ * @template Value
+ * @param {(name: string) => Value} mapper
+ * @returns {{ [name: string]: Value }}
```
More unsolicited feedback:

This is the first time I see the `@template` tag. I checked and it's not mentioned at [JSDoc.app](https://jsdoc.app/).

But then I found the [`@template` tag in TypeScript's docs](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template): it's for declaring generic functions, classes and types.

I would thus recommend renaming `Value` to `T` or `TValue` to clarify that `Values` is generic. It's a common convention to use `T` as the generic parameter name or to prefix generic parameter names with `T`.

---

```suggestion
 * @template T
 * @param {(name: string) => T} mapper
 * @returns {{ [name: string]: T }}
```

or

```suggestion
 * @template TValue
 * @param {(name: string) => TValue} mapper
 * @returns {{ [name: string]: TValue }}
```

(I would personally choose `T` in this case since I don't think that the "Value" in `TValue` brings any extra clarity.)

texastoland replied:
In general single letter names are bad practice even for generics. I could point to examples even within the compiler. In this specific case I agree `Value` doesn't add enough context to merit being different. Accepting option 1 👌🏼

mtsknn replied:
> In general single letter names are bad practice even for generics.

I agree, at least if there are more than one generic parameter. `T`, `U`, `R` etc. are confusing. `TFoo`, `TBar`, `TResult` etc. are better.

> I could point to examples even within the compiler.

I would be interested if you have time! Especially if you have examples with only one generic parameter where a single-letter name is bad.

> Accepting option 1 👌🏼

Good choice. 😉

chasefleming commented:
Hi @texastoland are you still working on the other changes you mentioned?

texastoland commented:
Yep finally pulled it down last night and wrapping up!

texastoland commented:
The following may not work as expected:

```js
const Integers = Counter()

const { Zero, One } = Integers // => 0, 1
const { ZERO, ONE } = Integers // => 2, 3
```

Still worth it?

chasefleming commented:
@texastoland I think that same issue exists currently if you use the reference instead of instantiating a new `Integers()`. We can create an issue to address that in a later version.

texastoland commented:
I was replying and realized just renaming them:

```js
const Integers /* doesn't imply which ones */ = Counter() /* state is intrinsic */
```

... resolves the issue from my perspective.

texastoland commented:
I got stuck on ezolenko/rollup-plugin-typescript2#304 but rebasing and finishing now 🏁

mtsknn commented on jsconfig.build.json:
```diff
@@ -0,0 +1,10 @@
+// non-default config name to quiet errors resovled by Microbundle
```
```suggestion
// non-default config name to quiet errors resolved by Microbundle
```

texastoland commented:
@chasefleming I'm ready for feedback before testing and documentation. The checklist in the description links to individual commits.

1 weird thing I noticed is `Integers` doesn't need a `Map` unless you expect the same values from the same keys every time. That seems reasonable so I made that behavior dead simple and applied it to `Symbols` too.

JSDoc types makes the code look more complex than it is. I'll show an example changed to TypeScript in case you're open to it. It already generates typings for TypeScript regardless.

With JSDoc (304 characters/11 lines):

```js
/**
 * @template T
 * @param {(name: string) => T} mapper always returns the same value for the same name
 */
export const MemoEnumOf = (mapper) =>
  EnumOf(
    (name, map) =>
      map.get(name) ??
      /** @type {!T} */ (map.set(name, mapper(name)).get(name)),
    /** @type {Map<string, T>} */ (new Map())
  )
```

With TypeScript (249 characters/8 lines):

```js
/**
 * @param mapper always returns the same value for the same name
 */
export const MemoEnumOf = <T>(mapper: (name: string) => T) =>
  EnumOf(
    (name, map) => map.get(name) ?? map.set(name, mapper(name)).get(name)!,
    new Map() as Map<string, T>
  )
```

Just JS (although the types already helped catch my own mistakes):

```js (206 characters/8 lines)
/**
 * @param mapper always returns the same value for the same name
 */
export const MemoEnumOf = (mapper) =>
  EnumOf(
    (name, map) => map.get(name) ?? map.set(name, mapper(name)).get(name),
    new Map()
  )
```

PS more than happy to explain anything odd!

chasefleming commented:
Thanks @texastoland. Will review as soon as I can. And re TypeScript, this library is purposefully not in TypeScript since enums are a feature there. Feels wrong to write it in TS 😊  It's probably unlikely people would use this if they are using TypeScript (Although, I'm curious to hear a counter to that with the new use cases you thought up. I think if we think up more use cases that could totally change.). But if it is the case that only JS codebases might use this library then the JSDocs are very nice for intellisense, but types files may not even be 100% necessary and just JS could be fine. Or could you just define variables for the arguments with types and then pass them in to make it look cleaner?

texastoland commented:
> if it is the case that only JS codebases might use this library then the JSDocs are very nice for intellisense

I use TypeScript (and ReScript etc.) because it catches my own errors including several times during this PR. In that sense using types affects writing code more than using it. In this case typings don't help autocomplete in JavaScript since the enum destructures to basically anything (there's nothing to autocomplete).

> not in TypeScript since enums are a feature there

I'm surprised `enum` (as well as `namespace`) was never officially deprecated. It's gently _discouraged_ in the official handbook and **unsupported** by Babel (and SWC etc.). The reasons are:

1. It relies on extra code at runtime:

```ts
enum Ordinals { A = 1, B, C, D }

// effectively generates the corresponding object

const Ordinals = {
  A:  1,  B:  2,  C:  3,  D:  4,
  1: "A", 2: "B", 3: "C", 4: "D",
}

// COMPARED TO

const { A, B, C, D } = Counter(1)
```

2. `enum` (unlike [Flow](https://flow.org/en/docs/enums/)) doesn't automatically generate strings:

```ts
enum Seasons {
  Summer = "Summer",
  Autumn = "Autumn",
  Winter = "Winter",
  Spring = "Spring",
}

// not better than

const Seasons = {
  Summer: "Summer",
  Autumn: "Autumn",
  Winter: "Winter",
  Spring: "Spring",
} as const

// COMPARED TO

const { Summer, Autumn, Winter, Spring } = Strings
```

3. Unions entirely replace `enum` for strings anyway:

```ts
type seasons = "Summer" | "Autumn" | "Winter" | "Spring"

function getSeasonalImage(season: seasons) {}

// type checks
getSeasonalImage("Summer")
// static error
getSeasonalImage("Autum")

// COMPARED TO

const { Summer, Autumn, Winter, Spring } = Strings

getSeasonalImage(Summer)
// runtime error without TypeScript (needs a unit test)
getSeasonalImage(Autum)
```

> a counter to that with the new use cases you thought up

1. `Counter` is the only API that overlaps `enum` and still necessary to compile without tsc (including every bundler).
2. `Strings` is still a relevant style choice compared to string literal unions.
3. Anything more creative (like `LowerCased` and `ActionCreators`) is beyond the scope of `enum`.

> could you just define variables for the arguments with types

Technically I could write the `.d.ts`s by hand but would lose type checking internally.

> to make it look cleaner

In retrospect it doesn't look bad to me. I added line and character counts to previous https://github.com/chasefleming/enum-xyz/pull/1#issuecomment-1066129808 to illustrate the difference in an intentionally extreme example. I'll follow your lead regardless ✌🏼

chasefleming commented:
Very interesting! I'm sold on use cases in TS. And yeah I don't think it actually looks that bad in JSDoc. This PR looks great. Excited to get this merged in. Awesome contribution.

chasefleming commented:
@texastoland Very interesting! I'm sold on use cases in TS. And yeah I don't think it actually looks that bad in JSDoc. This PR looks great. Excited to get this merged in. Awesome contribution.

chasefleming commented:
@texastoland Very interesting! I'm sold on use cases in TS. And yeah I don't think it actually looks that bad in JSDoc. This PR looks great. Excited to get this merged in. Awesome contribution. I see a few todo items still listed...are those still coming?

chasefleming commented on src/deprecated.js:
```diff
@@ -0,0 +1,49 @@
+ * @returns {import('./enum-xyz.js').Enum<T>}
+ */
+const deprecated = ({ from, to, using }) =>
+  EnumOf(
```
This could also be a cool example in the example use cases folder where the ActionCreators is going. I did something similar to this with in another project recently, but I like this a lot.

texastoland commented:
Yeah I just wanted to make sure you're happy with the code before I finish stuff depending on it!

chasefleming commented:
Yeah looks awesome @texastoland ...very happy! Also, definitely think adding that deprecation in the examples section would be nice as well so that we have more than one other use case. Very excited for this merge!

texastoland commented:
Documenting an issue I noticed in my previous example:

```ts
type seasons = "Summer" | "Autumn" | "Winter" | "Spring"

function getSeasonalImage(season: seasons) {}

// ELSEWHERE

const { Summer, Autumn, Winter, Spring } = Strings

// does NOOOT type check
getSeasonalImage(Summer)
// static error
getSeasonalImage(Autum)
```

I expected the type of `Summer` to be the string literal type `"Summer"`. It doesn't work because all the functions are declared using `string` for names. I tried changing to a generic type `TKey extends string` everywhere. That basically means:

> Each time the function is called it can be a different type but it has to be more specific than `string` (like `"Summer"`).

That didn't work either though:

> Only functions can leave types generic (as opposed to concrete like `string`) but we're using a proxy (object). Moreover destructuring must happen after any function call. TypeScript never looks "forward" for type information. Therefore nothing more could be inferred for a function either.

In conclusion TypeScript will still need a type annotation for `Strings` (as opposed to inferring it automatically):

```ts
const { Summer, Autum, Winter, Spring } = Strings as Strings<seasons>
//              ^^^^^
//              static error

// type checks
getSeasonalImage(Summer)
// static error
getSeasonalImage(Autum)
```

chasefleming commented:
@texastoland, yeah I see a lot of open GitHub issues around using proxies in TypeScript (specifically around inferring types). Looks like they're playing catch up. Could have another section as well for TypeScript tips with the library linked off of the main readme.

chasefleming commented:
By the way @texastoland , I added a CHANGELOG file if you could add a description of the updates to that please.

mtsknn commented:
An alternative approach:

```ts
const Seasons = Strings<'Summer' | 'Autumn' | 'Winter' | 'Spring'>()

function getSeasonalImage(season: keyof typeof Seasons) {}

getSeasonalImage(Seasons.Autumn) // ok
getSeasonalImage(Seasons.Autum) // error, typo

getSeasonalImage('Autumn') // ok
getSeasonalImage('Autum') // error, typo

// can also destructure but maybe not necessary:
const { Summer, Autumn, Winter, Spring, Foo } = Seasons
//                                      ^^^ error

getSeasonalImage(Autumn) // ok
getSeasonalImage(Autum) // error, typo
```

[TS Playground](https://www.typescriptlang.org/play?target=99#code/GYVwdgxgLglg9mABAUTCAtgHgCoD4AUAUIougIYAOFApgE4Bci+A1tQJ6MDOUtMYA5gBpE3MlGoB+RgFlKmbrwHC8ASkQBeXImyDiIqGMky5CvkO25CagN57a1KCFpIw1AO6IACrTgAPNkQkJNYAvrpBiNaI-A6M+AD6BrQxUMIUPjS0UGxqmqSUmfgAyjxm+OlwmdkqwqLiaiF6KoSNhKCQsAiI0tTocKgYOATkVHRxrBz6ivy5Wtg2dg5OSAPogUEs7MIjs-kUAIQAdCmbOYgSEntHnA6n2wV0pyoqx7cTKvvhQa4eshT4ajInG6JlKSgsTRahDa4Gg8CQJWmnBwiGovnEYAAJsDTAICAsSPZHM4UGg1k8NFp3oggZFEABtADSiD42gAuoxmY1Wu04V0ADJwNx0ADCQOoiLMyOwqPR1CxOLB-HxkUWxJWZPwFLyE0OUDgguFtDFNwBgOBUSZLKQ2A5iENdAg4swjK03OhvM6SAAchgAEZ0aWyjHYqZmAiiLIASSxaI0iAADATEETlt1ev1NQDKfoyNHY74ANSF810q2s22MMkB2iId2ECAIbiIIrUIFN+OSgTIgDkRQw6DoPcQAB9ED2AIIgRzoMDDsc9gDqfHEtHn46K6TMPfxHthXuiDlb7bAZAANlHyDF8DcT4wJnBgIhsjRHy225wmzZWilj5-TxeV7UMUH5NocU4zmAagAPTQYgcDMIQv6gQBl5kNef5gRBGAwXBL5wNCyEnueaHXpO04YHOuHwYhRH-iRQH4ORM49tR+HQrBiBOkg56fogmLUAoIDQE41D0A2TZQHS-boIOtDCNhs7CMuYCrsIm7TMIABicBwHWnYoZwhCcREplmeZFnmQAejZqK0D4tCEUeKEMehwGKVBiCcQhxlwXRCCudeilsWwFAEYQL7UIgIpwGecC1uo44AErUJi649gA4vY8rpQAQmeIDUD2ElgM2MVxbQwKJQ6xril2-DIuV8W7jCHTwoeUBNbQ+CNhVjCRW+XWcPSD5PgNT5DWy35OZ1sXxfgQ2HClmLUT5KRdQtc2VUt1AhWFM0bT29hpattEOIdy2sV5eGhXAtQABZwCAZ6YogAaIHFRpOjcSHnVtTH2FdnHsSVzZFPdlCCfGvroDWvYijAtAQGeRWjsl1DQGQAgo+l2C8Fj-A47ujalbF1CHHF-DFBDNCcIcCNIyj1EJsIPmcfYmPY1FnDMDAoyYqDZMU3AVPg5DdN4zABNM9diAAIzbDA-D3VJbhY1J+pvVFABMWtOiANwslAPbAlA91RWeQJSTA4joNaz7m4g4DtWQfpwAAbtQgso8Los04J9OI8ju2yyzNHQiTzZREUACOIB5tQwhJY9sMG-piUw3DmB9nHCfpcncCp5wO74HLSbe+TlP4AXRfUXLct62QBtRc3r0CUJImaeHkdC1Xsfx-Yddhz5QA)

But this would probably be useful only for TS users (though JS users could use destructuring). And numeric enums could have wrong order without destructuring (example in the TS Playground).

And eh 🤦‍♂️ I just realized that this is probably unnecessary because as @texastoland said:

> Unions entirely replace `enum` for strings anyway

But whatever, I had fun thinking about this. 🙃

texastoland commented:
@mtsknn FWIW I added this:

```ts
function getSeasonalImage(season: "summer" | "autumn" | "winter" | "spring") {}

// ELSEWHERE

const { Summer, Autum, Winter, Spring } = Lowercased as Lowercased<
//              ^^^^^
//              static error
  "Summer" | "Autumn" | "Winter" | "Spring"
>

// type checks
getSeasonalImage(Summer)
// static error
getSeasonalImage(Autum)
```

> Or could you just define variables for the arguments with types and then pass them in to make it look cleaner?

@chasefleming Moved them to a separate `types.ts` in 3460753dea5fbaabf420a8a15f9c6efb8388a5a9.

> I added a CHANGELOG file if you could add a description of the updates to that please.

Added a todo 🚀

chasefleming commented:
Cool, hope you don't mind I added "Add deprecated example in other use cases section" as well to the todos.

chasefleming commented:
@texastoland Cool, hope you don't mind I added "Add deprecated example in other use cases section" as well to the todos.

chasefleming commented on src/index.js:
```diff
@@ -0,0 +1,6 @@
+export * from './enum-xyz.js'
+export * from './deprecated.js'
```
Does this need to be exposed? In terms of other use cases, I was thinking we'd more just make a markdown file for each example use case we think up with a few words about them and then link them off of the readme.

chasefleming commented on src/index.js:
```diff
@@ -0,0 +1,6 @@
+export * from './enum-xyz.js'
+export * from './deprecated.js'
```
@texastoland Does this need to be exposed? In terms of other use cases, I was thinking we'd more just make a markdown file for each example use case we think up with a few words about them and then link them off of the readme.

texastoland replied:
If `deprecated` isn't exposed it's a breaking change in which case no need to have it. Or I misunderstand?

texastoland replied:
I think the Markdown is referring to something else right?

texastoland replied:
Resolving to continue in the main thread in context of `ActionCreators` too 👇🏼

texastoland commented:
@chasefleming Expanded from your review thread:

> Does this need to be exposed?

If `deprecated.js` isn't exposed it's a breaking change. In that case no need to have it?

> we should probably keep the existing functionality by referencing the new and add a deprecation warning with `console.warn` until they are removed in a later version.

Or did I misunderstand?

> In terms of other use cases, I was thinking we'd more just make a markdown file for each example use case we think up with a few words about them and then link them off of the readme. (e.g. Making a deprecator with `enumOf`)

I'm down for that. Or just include a heading in the readme? Can we make a separate docs PR together?

I added the Redux function in an `extras` module that isn't exported by default. There were several popular npm packages prior to the official `createAction`. I want to use it as an example too but without unnecessary implementation details.

chasefleming commented:
Oh sorry @texastoland you're totally right. I was thinking not to expose the actual `deprecated` function but it's hidden inside of the deprecated file and just the old methods are exposed. Please ignore my previous comment. Long day :)

And the extras stuff is great. Yeah let's do any docs stuff separate. You've done a lot! Let's get this in :)

chasefleming commented on src/enum-xyz.js:
```diff
@@ -0,0 +1,32 @@
+ * @type {import("./types").enumOf}
+ * @param mapper should return the same value for the same name
+ */
+export const enumOf = (mapper, state) =>
```
Very excited about this now being broken out as it's own!

chasefleming commented on src/enum-xyz.js:
```diff
@@ -0,0 +1,32 @@
+ * @type {import("./types").memoEnumOf}
+ * @param mapper always returns the same value for the same name
+ */
+export const memoEnumOf = (mapper) =>
```
Same with this!

texastoland replied:
However it's guaranteed here. A great example is `Symbol("X"`) != Symbol("X")`. This makes sure `Symbols.X === Symbols.X`!

texastoland commented on src/types.ts:
```diff
@@ -0,0 +1,29 @@
+// index.js
```
PS happy to do a video walkthrough of all changes including types and configs 🤘🏼

mtsknn replied:
I have no mental energy left to review this now but looks good at a very quick glance. 😄

texastoland commented:
Quick design question I was reminded of by @rauschma's [article](https://2ality.com/2022/01/symbol-factory.html#wouldn%E2%80%99t-it-be-better-to-cache-the-results?): when should these be memoized? It boils down to how are they intended to be used?

Conventional enums don't need memoization:

```js
// seasons.js
export const { Summer, Autumn, Winter, Spring } = Symbols

// another module
import { Summer, Autumn, Winter, Spring } from "seasons.js"

// yet another module
const { Summer, Autumn, Winnie, Brooke } = Symbols
// Summer1 !== Summer2
```

Whereas proxy-fied enums don't necessarily need exports:

```js
// import { Summer, Autumn, Winter, Spring } from "seasons.js"
// ... is technically more typing than ...
const { Summer, Autumn, Winter, Spring } = Symbols

// another module
const { Summer, Autumn, Winnie, Brooke } = Symbols
// with the notable caveat Summer1 === Summer2
```

The first option helps prevent typos whereas the second just saves typing quotes (and calling `Symbol.for`). The `type` would also need imported in the second case regardless as concluded in https://github.com/chasefleming/enum-xyz/pull/1#issuecomment-1069706711.

An API solution might be to provide either option depending whether it's used as a function or not:

```js
const { Summer, Autumn, Winter, Spring } = Symbols

// another module
const { Summer, Autumn, Winter, Spring } = Symbols
// where Summer1 === Summer2

// yet another module
const { Summer, Autumn, Winnie, Brooke } = Symbols()
// Summer2 !== Summer3
```

The intuition would be that the function version acts provides a distinct "scope" similar to functions having a distinct `this`.

Equivalently for numbers:

```js
const { A, B, C, D } = Integers // 0, 1, 2, 3

// another module
const { A, B, C, D } = Integers // 4, 5, 6, 7
// where A1 === A2

// no need for Counter anymore
const { A, B, E, F } = Integers(1) // 1, 2, 3, 4
// A2 !== A3
// B2 !== A3
```

chasefleming commented:
@texastoland Yeah it seems like there are two use cases we should be intentional about and I like having the option for both. The colors and mood example from the article is a good one:

```javascript
// EXAMPLE 1

// colors
const {red, green, blue} = Strings

// mood
const {happy, blue} = Strings

blue === blue // should be false
```

In that case, blue should not be the same. Symbol is good here. But there is a use case where someone may have a string from a JSON response they need to check against the enum (for example):

```javascript
// EXAMPLE 2

const { v1 } = Strings

v1 === "v1" // we want to be true
```

So correct me if I'm wrong but what you are proposing is to make "Example 1" use `Strings()` instead of `Strings` as an indication to use Symbol and scope it.

Which I think is actually better than TypeScript enums, because if I'm not mistaken the two "Blues" in the following example evaluate to true:

```
enum mood {
  Blue,
  Happy
}

enum colors {
  Blue,
  Yellow
}

mood.Blue === colors.Blue // should not be true but is in TypeScript
```

I'm very into the optional scoping, but technically, how would you go about making that scoping options though? You'd have to return an object or an IIFE then how would you invoke another function on top of that? Not sure you could curry an IIFE since it doesn't return a function(or an optional one at that).

But I also think it may cause confusion to a developer. It's a pretty subtle difference and I think being more explicit is probably prone to less issues for developers. So something like this would maybe be clearer.

```
const { blue, happy } = Strings({ unique: true })

// or

const { blue, happy } = Strings().unique()

// or 

const { blue, happy } = Strings().isUnique()
```

But that loses the simplicity in my opinion and starts to get pretty ugly and lengthy...which is one of the reasons I went with the original design. Maybe you have a suggestions that is more explicit design but simpler? Although if it's a rarer case that you'd want it scoped it's not terrible to have the additional method and my vote would be for `.unique()`. Maybe `StringsLower` would also be replaced `Strings().lower()` in this case? Then could chain unique if you wanted.

(If we do think up other options though a case can be made for other options then a function approach would probably make more sense anyway)

texastoland commented:
Moving tests and PR checks todos to separate PRs as suggested on Discord.

chasefleming commented:
@texastoland let's do the colors item as a separate PR as well so we can get this in. Should probably have tests for the new method namings at least, but should be able to copy those easily from the existing tests. Think changelog should be in this one as well under "Unreleased"

texastoland commented:
Tests passing but 1 small typing error to fix 🐛

chasefleming commented on src/types.ts:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @example
```
nice touch!

texastoland commented:
The diff is confusing even for me. We have the conversation here for context. Opening sequential PRs for more granular changes ASAP!
